Methods Section Review (#5)
erickabsmith/masters-project-lcz-classification: doc/report.Rmd

```

## Random Forests

> Random forests consist of many decision trees. A decision tree can be used for categorization or regression. Here I’ll focus on categorization since my goal is to predict LCZ class, which is a categorical variable. Decision trees work by putting each observation in the original (“training”) dataset through a series of conditional statements, or tests, in order to group it with other observations that are similar. These similar groups are expected to have similar values for the variable you’re interested in predicting. Since the true value of that variable is known for the observations in the training dataset it’s possible to measure the accuracy of each prospective structuring of these conditional statements, and therefore to pick the best tree. 

Author: ```suggestion
Random forests consist of many decision trees. A decision tree can be used for classification or regression. Here I’ll focus on classification since my goal is to predict LCZ class, which is a categorical variable. Decision trees work by putting each observation in the original (“training”) dataset through a series of conditional statements, or tests, in order to group it with other observations that are similar. These similar groups are expected to have similar values for the variable you’re interested in predicting. Since the true value of that variable is known for the observations in the training dataset it’s possible to measure the accuracy of each prospective structuring of these conditional statements, and therefore to pick the best tree. 
```

Splits are typically evaluated by Gini impurity or entropy.

$$
\text{Gini Impurity} =\ I_G(t)\  = 1 - \sum_{i=1}^{C}p(i|t)^2
$$

$$
\text{Entropy} =\ I_H(t)\ = -\sum_{i=1}^{C}p(i|t)log_2p(i|t)

> $$

Author: Use `\text{}` around any full words in your equations - otherwise LaTeX typesets each letter as though it is a variable, and the spacing gets wacky.  I've fixed it here, but you need to double check your other equations and make the same fix.

Author: Also add a backslash in front of log, e.g. `\log_2{p(i|t)}` to get it typeset a bit nicer too.

> Where $p(i|t)$ is the proportion of samples that belong to class $i$ for a particular node $t$.

Author: You also need to define C.

> A Gini Impurity of 0 or an entropy value of 0 indicates a completely homogeneous group, which cannot be improved upon. A split doesn’t need to have a value of 0 to create the best node though, these metrics are used for comparison. All possible different variables and different cutoffs within those variables are all tried and their Gini impurity or entropy calculated, then, the one with the best (lowest) value is the split used at that node. This is done recursively at each node until all possible splits no longer offer a decrease in the metric of choice. The resulting collection of nodes is a decision tree. 

Author: An attempt to make this paragraph more concise:
```suggestion
A Gini Impurity, or entropy, of 0 indicates a completely homogeneous group, which cannot be improved upon. These metrics are used for comparison: all possible variables and thresholds within those variables are tried and their Gini impurity or entropy calculated, the variable and threshold with the best (lowest) value selected for the node. This is done recursively at each node until no split offers a decrease in the metric of choice. The resulting collection of nodes is a decision tree. 
```

A Gini Impurity of 0 or an entropy value of 0 indicates a completely homogeneous group, which cannot be improved upon. A split doesn’t need to have a value of 0 to create the best node though, these metrics are used for comparison. All possible different variables and different cutoffs within those variables are all tried and their Gini impurity or entropy calculated, then, the one with the best (lowest) value is the split used at that node. This is done recursively at each node until all possible splits no longer offer a decrease in the metric of choice. The resulting collection of nodes is a decision tree. 

Decision trees perform quite poorly with new samples. Including a threshold value for the accuracy metric can help, as it keeps trees more simple, but they are still prone to overfitting. Collecting them into a random forest addresses this issue. True to its name, a random forest is a collection of decision trees that have a component of randomness in their creation. The predictive aspect of a random forest doesn’t only rely on one tree, it is determined by totaling up the decisions, or “votes,” that all of the trees make.

> To create each tree the first step is bootstrapping the training dataset. The decision tree is then created with that bootstrapped data, but with slightly different rules. For each node, only a subset of the variables are randomly selected and used as candidates to split up the bootstrapped data. Not all the variables are used at each node because doing so reduces correlation and introduces randomness into the model. Whichever variable best splits the data will be the variable that is kept at that node in that specific tree. This is done recursively until no more splits are beneficial, just as in a regular decision tree. To create the next tree the process is the same, but starts completely over with a new bootstrap sample of the training data. This is repeated for a chosen number of trees.

Author: "This is done recursively until no more splits are beneficial, just as in a regular decision tree."  True, only if minimum size is not reached.

> This choice of the number of trees to create is a tuning parameter. Too many trees can be computationally expensive, but too few can create a model that that is poor at prediction. Another important tuning parameter is the choice of how many variables to randomly select at each node in the trees must be made. The default value for this is typically the number of variables squared for classification or the number of variables divided by 3 for regression. A less common, but useful, way to tweak the model is by trying different values for the minimum size of terminal nodes. A large minimum size means smaller trees and faster completion of the random forest. In addition, the maximum number of terminal nodes can be set to a specific number. Otherwise trees are grown as large as possible. Depth, which indicates the number of splits a tree has, can also be controlled. Greater depth gives more information about the data, but can cause overfitting.  

Author: Isn't is "square root of the number of variables"? Not squared.

> This choice of the number of trees to create is a tuning parameter. Too many trees can be computationally expensive, but too few can create a model that that is poor at prediction. Another important tuning parameter is the choice of how many variables to randomly select at each node in the trees must be made. The default value for this is typically the number of variables squared for classification or the number of variables divided by 3 for regression. A less common, but useful, way to tweak the model is by trying different values for the minimum size of terminal nodes. A large minimum size means smaller trees and faster completion of the random forest. In addition, the maximum number of terminal nodes can be set to a specific number. Otherwise trees are grown as large as possible. Depth, which indicates the number of splits a tree has, can also be controlled. Greater depth gives more information about the data, but can cause overfitting.  

Author: ```suggestion
The hoice of the number of trees to create is a tuning parameter. Too many trees can be computationally expensive, but too few can create a model that that is poor at prediction. Another important tuning parameter is the number variables to randomly select at each node in the trees. The default value for this is typically the number of variables squared for classification or the number of variables divided by 3 for regression. A less common, but useful, way to tweak the model is by trying different values for the minimum size of terminal nodes. A large minimum size means smaller trees and faster completion of the random forest. In addition, the maximum number of terminal nodes can be set to a specific number. Otherwise trees are grown as large as possible. Depth, which indicates the number of splits a tree has, is another way to control tree size. Greater depth may improve the performance of individual trees, but can cause overfitting.  
```

> Once the model has been tuned to optimality, the random forest is used to decide about a new set of input values. This is called bagging, because you’re bootstrapping the data and using the aggregate to make a decision . Bagging is useful because it reduces variance without introducing bias. The decision itself is made by putting the new set of input values into each decision tree individually and seeing which outcome each tree classifies the variable of interest into. These are all combined either by direct vote (for categorical variables) or an average (for quantitative variables) and this becomes the final predicted result.

Author: I think make this a stand alone paragraph about using a tree to make a prediction, e.g. 
```suggestion
To use a random forest to make predictions for a new set of input values, the values are fed into each decision tree individually. The predictions from each tree are combined either by direct vote (for categorical variables) or an average (for quantitative variables) and this becomes the final predicted result. This process is called bagging, because you’re bootstrapping the data to create each tree and using the aggregate to make a decision. Bagging is useful because it reduces variance without introducing bias. 
```
